### ISSUE TITLE
Field.endsWith() and similar methods should escape argument lazily

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#8429]
Our convenience methods `Field.endsWith()` use the `LIKE` predicate behind the scenes. In order to transparently implement `.. LIKE .. ESCAPE ..` in case users use `%` or `_` wildcards with this method, we escape their arguments using `Tools.escapeForLike()`.

Unfortunately, this is being done eagerly when `endsWith()` is called:

```java
    public final Condition endsWith(T value) {
        if (value instanceof Field)
            return endsWith((Field) value);

        Field<String> concat = DSL.concat(inline("%"), Tools.escapeForLike(value));
        return like(concat, Tools.ESCAPE);
    }
```

As users may decide to replace their actual bind variables only later on, the re-escaping won't work and they get wrong results, see this issue here: #8422.

We should maintain the user bind variable until we generate the SQL statement, and escape it either directly in SQL (most reliable in the presence of user defined data type bindings), or prior to binding it (less reliable).

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.12. Backport scheduled for 3.11.12 (#8525)

**Comment by lukaseder** [comment]
Backport to 3.11.12 is too risky due to merge conflicts. Since the user has a workaround and this is central functionality, we'll publish this fix with 3.12 only (soon)
